cunumeric/module.py

```diff
@@ -1257,6 +1278,44 @@ def check_shape_dtype(
     return converted, ArrayInfo(ndim, shape, dtype)
 
 
+def _block(arr, cur_depth, depth):
```

Comment:
I'm afraid that this function is also too convoluted for anyone to verify correctness. Why can't we just use the original `arrays` to do this? Here's what NumPy's documentation says `np.block` does:

> Blocks in the innermost lists are concatenate along the last dimension (-1), then these are concatenated along the second-last dimension (-2), and so on until the outermost list is reached.

So, similar to my rewrite of the checker, I guess you can traverse down until you hit a list of non-list elements whose depth is 1, concatenate them with this depth, and pop back up and repeat the same. I don't see why this involves any bookkeeping of siblings like you do in this code, unless you convince me they that's absolutely necessary.

Reply:
Actually the bookeeping is to notify where the misalignment happens. Numpy raises value error a bit more precisely. 
I can delete that and make it simple as you said.

Reply:
I'm not necessarily suggesting that we should delete it. I just don't want it to be tangled up with the input, which made the previous code hard to crack.

Reply:
> Actually the bookeeping is to notify where the misalignment happens. Numpy raises value error a bit more precisely.
are you planning to implement this?

Reply:
No, I think the current implementation provides enough information to let users know where the misalignment happens.